app/src/main/kotlin/compiler/intermediate_form/Linearization.kt

```diff
@@ -1,10 +1,111 @@
         cfg: ControlFlowGraph,
         covering: Covering,
-    ): List<Asmable> = TODO()
+    ): List<Asmable> {
+        if (cfg.entryTreeRoot == null)
+            return emptyList()
+
+        val instructions = mutableListOf<Asmable>()
+        val labels = referenceHashMapOf<IFTNode, String>()
+        val endLabel = "end"
+        val usedLabels = mutableSetOf<String>()
+
+        fun assignLabel(node: IFTNode): String {
+            val label = "_" + labels.size
+            labels[node] = label
+            return label
+        }
+
+        fun addLabel(label: String) {
+            instructions.add(Label(label))
+        }
+
+        fun dfs(node: IFTNode, nextLabel: String) {
+            addLabel(labels[node] ?: assignLabel(node))
+
+            fun addUncoditional() {
+                instructions.addAll(covering.coverUnconditional(node))
+            }
+
+            fun addConditional(whenTrue: Boolean, targetLabel: String) {
+                instructions.addAll(covering.coverConditional(node, targetLabel, !whenTrue))
+                usedLabels.add(targetLabel)
+            }
+
+            fun addJump(targetLabel: String) {
+                if (targetLabel != nextLabel) {
+                    instructions.add(Jmp(targetLabel))
+                    usedLabels.add(targetLabel)
+                }
+            }
+
+            if (node in cfg.unconditionalLinks) {
+                val target = cfg.unconditionalLinks[node]!!
+
+                addUncoditional()
+
+                if (target in labels)
+                    addJump(labels[target]!!)
+                else
+                    dfs(target, nextLabel)
+            } else if (node in cfg.conditionalTrueLinks && node in cfg.conditionalFalseLinks) {
+                val targetWhenTrue = cfg.conditionalTrueLinks[node]!!
+                val targetWhenFalse = cfg.conditionalFalseLinks[node]!!
+
+                if (targetWhenTrue in labels) {
+                    addConditional(true, labels[targetWhenTrue]!!)
+
+                    if (targetWhenFalse in labels)
+                        addJump(labels[targetWhenFalse]!!)
+                    else
+                        dfs(targetWhenFalse, nextLabel)
+                } else if (targetWhenFalse in labels) {
+                    addConditional(false, labels[targetWhenFalse]!!)
+                    dfs(targetWhenTrue, nextLabel)
+                } else {
+                    val label = assignLabel(targetWhenFalse)
+
+                    addConditional(false, label)
+
+                    dfs(targetWhenTrue, label)
+                    dfs(targetWhenFalse, nextLabel)
+                }
+            } else if (node in cfg.conditionalTrueLinks) {
+                val target = cfg.conditionalTrueLinks[node]!!
+
+                if (target in labels) {
+                    addConditional(true, labels[target]!!)
+                    addJump(endLabel)
+                } else {
+                    addConditional(false, endLabel)
```

Comment:
If there is only a link for the `true` case, then in the `false` case the control flow reaches the end. The `nextLabel` is only used to skip unnecessary jumps (see how `addJump` works), and here a jump to `endLabel` is necessary, because it is followed by some sequence of instructions generated by `dfs` in the following line.

By the way, this situation will never occur in practice, because all CFGs generated for functions are suffixed by epilogues (`ControlFlow.attachPrologueAndEpilogue`), and so there won't be any node with only one conditional link (because the are no conditionals in the epilogue). This code, however, does not assume this, and covers this situation anyway. (And so it is possible to linearize the function's CFG and its epilogue's CFG separately.)

Reply:
Thanks for explanation, it's clear now.